# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1345491474 14400
#      Mon Aug 20 15:37:54 2012 -0400
# Node ID 2a36a5b89f9874a209b6c2d082fba85c43acb205
# Parent  2f2fb099ac3a7e91b193d2b140814324d73f1cef
link gui in way that is more consistent with the rest of Octave

* libgui/src/link-deps.mk: New file.
* libgui/src/Makefile.am: Include it.
(QT_INCDIR, QT_LIBDIR, QT_LIBS, QT_LDFLAGS): New variables.
(liboctgui_la_CPPFLAGS): Use $(QT_INCDIR).
(liboctgui_current, liboctgui_revision, liboctgui_age,
liboctgui_version): Ne variables.
(liboctgui_la_LDFLAGS): Use -version option, $(NO_UNDEFINED_LDFLAG),
-bindir option, and $(LIBOCTGUI_LINK_OPTS).
(liboctgui_la_LIBADD): Use $(LIBOCTGUI_LINK_DEPS).
* src/Makefile.am: Include ../libgui/src/link-deps.mk.
(octave_LDADD): Move $(OCTAVE_GUI_LIBS) ahead of $(OCTAVE_CORE_LIBS).
Use $(OCTAVE_GUI_LINK_DEPS).
(octave_LDFLAGS): Use $(NO_UNDEFINED_LDFLAG) and
$(OCTAVE_GUI_LINK_OPTS) instead of $(octave_cli_LDFLAGS).
* liboctave/link-deps.mk, libinterp/link-deps.mk: Use $(top_srcdir)/
instead of ../ to reference other link-deps.mk files.

diff --git a/libgui/src/Makefile.am b/libgui/src/Makefile.am
--- a/libgui/src/Makefile.am
+++ b/libgui/src/Makefile.am
@@ -13,16 +13,24 @@
 # for more details.
 #
 # You should have received a copy of the GNU General Public License
 # along with Octave; see the file COPYING.  If not, see
 # <http://www.gnu.org/licenses/>.
 
 include $(top_srcdir)/build-aux/common.mk
 
+QT_INCDIR = @QT_INCDIR@
+
+QT_LIBDIR = @QT_LIBDIR@
+
+QT_LIBS = -lQtCore -lQtGui -lQtNetwork -lqscintilla2
+
+QT_LDFLAGS = -L$(QT_LIBDIR)
+
 SRCDIRS = \
   m-editor \
   qtinfo \
   octave-adapter \
   ../qterminal/libqterminal/unix \
   ../qterminal/libqterminal/win32 \
   ../qterminal/libqterminal \
   ../qterminal
@@ -76,30 +84,23 @@ liboctgui_la_SOURCES = \
   resource-manager.cc \
   settings-dialog.cc \
   symbol-information.cc \
   terminal-dockwidget.cc \
   welcome-wizard.cc \
   workspace-model.cc \
   workspace-view.cc
 
-liboctgui_la_LIBADD = \
-  libqterminal.la \
-  -lQtCore \
-  -lQtGui \
-  -lQtNetwork \
-  -lqscintilla2
-
 nodist_liboctgui_la_SOURCES = $(octave_gui_MOC) $(octave_gui_RC)
 
 liboctgui_la_CPPFLAGS = \
-  -I@QT_INCDIR@ \
-  -I@QT_INCDIR@/QtCore \
-  -I@QT_INCDIR@/QtGui \
-  -I@QT_INCDIR@/QtNetwork \
+  -I$(QT_INCDIR) \
+  -I$(QT_INCDIR)/QtCore \
+  -I$(QT_INCDIR)/QtGui \
+  -I$(QT_INCDIR)/QtNetwork \
   -I$(srcdir)/../qterminal/libqterminal \
   -I$(srcdir)/m-editor \
   -I$(srcdir)/octave-adapter \
   -I$(srcdir)/qtinfo \
   -I$(srcdir)/../../libcruft/misc \
   -I../../liboctave \
   -I$(srcdir)/../../liboctave \
   -I../../libinterp \
@@ -108,17 +109,36 @@ liboctgui_la_CPPFLAGS = \
   -I../../libinterp/parse-tree \
   -I$(srcdir)/../../libinterp \
   -I$(srcdir)/../../libinterp/interp-core \
   -I$(srcdir)/../../libinterp/interpfcn \
   -I$(srcdir)/../../libinterp/octave-value \
   -I$(srcdir)/../../libtinerp/operators \
   -I$(srcdir)/../../libinterp/parse-tree
 
-liboctgui_la_LDFLAGS = -L@QT_LIBDIR@
+include link-deps.mk
+
+liboctgui_la_LIBADD = \
+  libqterminal.la \
+  ../../liboctave/liboctave.la \
+  ../../libcruft/libcruft.la \
+  $(LIBOCTGUI_LINK_DEPS)
+
+# Increment these as needed and according to the rules in the libtool manual:
+liboctgui_current = 0
+liboctgui_revision = 0
+liboctgui_age = 0
+
+liboctgui_version_info = $(liboctgui_current):$(liboctgui_revision):$(liboctgui_age)
+
+liboctgui_la_LDFLAGS = \
+  -version-info $(liboctgui_version_info) \
+  $(NO_UNDEFINED_LDFLAG) \
+  -bindir $(bindir) \
+  $(LIBOCTGUI_LINK_OPTS)
 
 noinst_HEADERS = \
   documentation-dockwidget.h \
   files-dockwidget.h \
   history-dockwidget.h \
   m-editor/file-editor-interface.h \
   m-editor/file-editor-tab.h \
   m-editor/file-editor.h \
@@ -219,20 +239,19 @@ noinst_HEADERS += \
   ../qterminal/libqterminal/unix/ScreenWindow.h \
   ../qterminal/libqterminal/unix/TerminalCharacterDecoder.h \
   ../qterminal/libqterminal/unix/Vt102Emulation.h \
   ../qterminal/libqterminal/unix/SelfListener.h \
   ../qterminal/libqterminal/unix/TerminalModel.h \
   ../qterminal/libqterminal/unix/TerminalView.h
 
 libqterminal_la_CPPFLAGS = \
-  $(DEFS) \
-  -I@QT_INCDIR@ \
-  -I@QT_INCDIR@/QtCore \
-  -I@QT_INCDIR@/QtGui \
+  -I$(QT_INCDIR) \
+  -I$(QT_INCDIR)/QtCore \
+  -I$(QT_INCDIR)/QtGui \
   -I$(srcdir)/../qterminal/libqterminal
 
 libqterminal_la_MOC = \
   ../qterminal/libqterminal/moc-QTerminal.cc \
   ../qterminal/libqterminal/moc-QTerminalInterface.cc
 
 nodist_libqterminal_la_SOURCES = $(libqterminal_la_MOC)
 
diff --git a/libgui/src/link-deps.mk b/libgui/src/link-deps.mk
new file mode 100644
--- /dev/null
+++ b/libgui/src/link-deps.mk
@@ -0,0 +1,21 @@
+include $(top_srcdir)/libinterp/link-deps.mk
+
+if AMCOND_ENABLE_DYNAMIC_LINKING
+  LIBOCTGUI_LINK_DEPS =
+else
+  LIBOCTGUI_LINK_DEPS = $(DLDFCN_LIBS)
+endif
+
+LIBOCTGUI_LINK_DEPS += \
+  $(QT_LIBS)
+
+LIBOCTGUI_LINK_OPTS = \
+  $(QT_LDFLAGS)
+
+if AMCOND_LINK_ALL_DEPS
+  LIBOCTGUI_LINK_DEPS += $(LIBOCTINTERP_LINK_DEPS)
+  LIBOCTGUI_LINK_OPTS += $(LIBOCTINTERP_LINK_OPTS)
+
+  OCTAVE_GUI_LINK_DEPS = $(LIBOCTGUI_LINK_DEPS)
+  OCTAVE_GUI_LINK_OPTS = $(LIBOCTGUI_LINK_OPTS)
+endif
diff --git a/libinterp/link-deps.mk b/libinterp/link-deps.mk
--- a/libinterp/link-deps.mk
+++ b/libinterp/link-deps.mk
@@ -1,9 +1,9 @@
-include ../liboctave/link-deps.mk
+include $(top_srcdir)/liboctave/link-deps.mk
 
 if AMCOND_ENABLE_DYNAMIC_LINKING
   LIBOCTINTERP_LINK_DEPS =
 else
   LIBOCTINTERP_LINK_DEPS = $(DLDFCN_LIBS)
 endif
 
 LIBOCTINTERP_LINK_DEPS += \
diff --git a/liboctave/link-deps.mk b/liboctave/link-deps.mk
--- a/liboctave/link-deps.mk
+++ b/liboctave/link-deps.mk
@@ -1,9 +1,9 @@
-include ../libcruft/link-deps.mk
+include $(top_srcdir)/libcruft/link-deps.mk
 
 LIBOCTAVE_LINK_DEPS = \
   $(SPARSE_XLIBS) \
   $(ARPACK_LIBS) \
   $(QRUPDATE_LIBS) \
   $(FFTW_XLIBS) \
   $(LAPACK_LIBS) \
   $(BLAS_LIBS) \
diff --git a/src/Makefile.am b/src/Makefile.am
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -61,16 +61,18 @@ EXTRA_DIST = \
   octave-config.in.cc \
   octave-config.in.sh
 
 OCTAVE_CORE_LIBS = \
   ../libinterp/liboctinterp.la \
   ../liboctave/liboctave.la \
   ../libcruft/libcruft.la
 
+include ../libgui/src/link-deps.mk
+
 octave_cli_SOURCES = main-cli.cc
 
 octave_cli_LDADD = \
   $(OCTAVE_CORE_LIBS)
   $(OCTAVE_LINK_DEPS)
 
 octave_cli_LDFLAGS = \
   $(NO_UNDEFINED_LDFLAG) \
@@ -86,21 +88,23 @@ else
   OCTAVE_GUI_CPPFLAGS =
 endif
 
 octave_CPPFLAGS = \
   $(AM_CPPFLAGS) \
   $(OCTAVE_GUI_CPPFLAGS)
 
 octave_LDADD = \
+  $(OCTAVE_GUI_LIBS) \
   $(OCTAVE_CORE_LIBS) \
-  $(OCTAVE_GUI_LIBS) \
-  $(OCTAVE_LINK_DEPS)
+  $(OCTAVE_GUI_LINK_DEPS)
 
-octave_LDFLAGS = $(octave_cli_LDFLAGS)
+octave_LDFLAGS = \
+  $(NO_UNDEFINED_LDFLAG) \
+  $(OCTAVE_GUI_LINK_OPTS)
 
 if AMCOND_BUILD_COMPILED_AUX_PROGRAMS
 octave-config.cc: octave-config.in.cc Makefile
 	@$(do_subst_default_vals)
 
 mkoctfile.cc: mkoctfile.in.cc Makefile
 	@$(do_subst_config_vals)
 else
